Product status is in stock on storefront even it is out of stock in all sources 2156

vpodorozh opened the pull request:
### Steps to reproduce (*)
1. Go to create product page
2. Fill all required data
3. Assign a few sources
4. Switch **Source Item Status** to **Out of Stock** on each source
5. Switch **Backorders** to **Allow Qty Below 0** on Advanced Inventory page
6. Save product
7. Open product on storefront

### Expected result (*)
Product isn't salable and status is **Out of Stock**

### Actual result (*)
Product is salable and status is **In Stock**

### Fixed Issues - #2156

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Added integration test for case when there are backorders along with out of stock items

vpodorozh committed:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Added fix.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2172) <br/>All committers have signed the CLA.

ishakhsuvarov commented:
@vpodorozh @novikor Great teamwork guys

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsAnySourceInStockCondition.php:
```diff
@@ -0,0 +1,49 @@
+     */
+    public function execute(string $sku, int $stockId): bool
+    {
+        $sourceItems = $this->getSourceItemsBySku->execute($sku);
```
It could make sense to use `\Magento\InventoryApi\Api\SourceItemRepositoryInterface::getList` with an appropriate search criteria here

vpodorozh replied:
Yeh. Done.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceInStockCondition.php:
```diff
@@ -0,0 +1,73 @@
+    private $productSalableResultFactory;
+
+    /**
+     * IsAnySourceInStockCondition constructor.
```
Yes

vpodorozh replied:
No, but yes :)

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceInStockCondition.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * @inheritdoc
+ */
+class IsAnySourceInStockCondition implements IsProductSalableForRequestedQtyInterface
```
Maybe something like `IsAnySourceItemInStock`?

vpodorozh replied:
Agree. renamed

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceInStockCondition.php:
```diff
@@ -0,0 +1,73 @@
+            $errors = [
+                $this->productSalabilityErrorFactory->create([
+                    'code' => 'stock_item_is_any_source_in_stock-no_sources_in_stock',
+                    'message' => __('There is no sources in stock')
```
It would be nice to clarify error message.
Also, please make sure that phrase is added to the translations csv file

vpodorozh replied:
Clarified message

Commit:
Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156.
Test condition instead of chain

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceInStockCondition.php:
```diff
@@ -0,0 +1,73 @@
+            return $this->productSalableResultFactory->create(['errors' => $errors]);
+        }
+
+        return $this->productSalableResultFactory->create(['errors' => []]);
```
You could as well add `$error = []` at the top and simplify returns

vpodorozh replied:
done

vpodorozh committed:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.

vpodorozh committed:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.

vpodorozh committed:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.

Commit:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Tests coverage IsProductSalableForRequestedQty

vpodorozh commented:
@ishakhsuvarov - I've made needed changes according to your review. Please review.

vpodorozh committed:
Merge branch '2.3-develop' into Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156

Commit:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Skip verification for non-manageable product types.

Commit:
Merge remote-tracking branch 'origin/Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156' into Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156

novikor committed:
Merge branch '2.3-develop' into Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156

ishakhsuvarov commented:
@vpodorozh @novikor Please check the failing integration tests

maghamed commented on app/code/Magento/InventoryApi/Test/_files/source_items_out_of_stock.php:
```diff
@@ -0,0 +1,89 @@
+/**
+ *  Copyright © Magento, Inc. All rights reserved.
+ *  See COPYING.txt for license details.
+ */
```
copyright block should be fixed:
```
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
```

extra spaces should be removed

VitaliyBoyko replied:
@maghamed you got shifty eyes :)

maghamed commented on app/code/Magento/InventorySales/etc/di.xml:
```diff
@@ -59,6 +59,10 @@
                     <item name="required" xsi:type="boolean">true</item>
                     <item name="object" xsi:type="object">Magento\InventorySales\Model\IsProductSalableCondition\IsSetInStockStatusForCompositeProductCondition</item>
                 </item>
+                <item name="stock_item_is_any_source_in_stock" xsi:type="array">
```
better to change name on `is_any_source_item_in_stock`

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,72 @@
+            $errors = [
+                $this->productSalabilityErrorFactory->create([
+                    'code' => 'stock_item_is_any_source_in_stock-no_source_items_in_stock',
+                    'message' => __('There are no source items with in stock status')
```
all new phrases should be added to translatibon file

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,72 @@
+use Magento\InventorySalesApi\Api\Data\ProductSalableResultInterfaceFactory;
+use Magento\InventorySalesApi\Api\Data\ProductSalabilityErrorInterfaceFactory;
+use Magento\InventorySales\Model\IsProductSalableCondition\IsAnySourceItemInStockCondition
+    as IsAnySourceInStockConditionCondition;
```
let's change the name, for this class as `IsAnySourceInStockConditionCondition` ConditionCondition - does not make a lot of sense . 
Better to remove condition suffix at all

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,92 @@
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+
+/**
+ * @inheritdoc
```
the individual description for this condition should be provided.

Describing What this condition is responsible for

maghamed commented on app/code/Magento/InventoryApi/Test/_files/source_items_out_of_stock.php:
```diff
@@ -0,0 +1,89 @@
+ *
+ * SKU-3 - EU-source-2(id:20) - 6qty (out of stock)
+ */
+$sourcesItemsData = [
```
guys, we already have dedicated fixture 
`app/code/Magento/InventoryApi/Test/_files/source_items.php`

let's extend existing one and make it fit for all cases, no need to introduce new fixture every time when we need to cover new business scenario, moreover copy-pasting 95% of all the existing code

maghamed commented on app/code/Magento/InventoryApi/Test/_files/source_items_out_of_stock_rollback.php:
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ *  Copyright © Magento, Inc. All rights reserved.
+ *  See COPYING.txt for license details.
```
again incorrect format of copyright

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,92 @@
+     */
+    public function execute(string $sku, int $stockId): bool
+    {
+        if (!$this->isProductStockManageable($sku)) {
```
Why do we have this check at all? We have a dedicated condition responsible for Manage Stock check - `\Magento\InventorySales\Model\IsProductSalableCondition\ManageStockCondition`

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,92 @@
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter(SourceItemInterface::SKU, $sku)
+            ->addFilter(SourceItemInterface::STATUS, SourceItemInterface::STATUS_IN_STOCK)
+            ->create();
```
This is incorrect because you check whether Product is In Stock on at least ONE Source, but instead, you need to check whether Product is In Stock on at least ONE Source `assigned to provided Stock`. You are missing the last part.
 

Thus, what you need to do is:
1. call `\Magento\InventoryApi\Api\GetSourcesAssignedToStockOrderedByPriorityInterface::execute($stockId)`
which returns `\Magento\InventoryApi\Api\Data\SourceInterface[]`
2. add source codes to the filter of your SourceItem Repository::getList search criteria

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,92 @@
+     *
+     * @return bool
+     */
+    private function isProductStockManageable(string $sku): bool
```
you no need this method in this class

maghamed commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsAnySourceItemInStockCondition.php:
```diff
@@ -0,0 +1,72 @@
+        if (!$isValid) {
+            $errors = [
+                $this->productSalabilityErrorFactory->create([
+                    'code' => 'stock_item_is_any_source_in_stock-no_source_items_in_stock',
```
make the code more user friendly

maghamed commented on app/code/Magento/InventorySales/Test/Integration/IsProductSalable/IsAnySourceItemInStockConditionTest.php:
```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ *  Copyright © Magento, Inc. All rights reserved.
+ *  See COPYING.txt for license details.
```
fix format

maghamed commented on app/code/Magento/InventorySales/Test/Integration/IsProductSalable/IsAnySourceItemInStockConditionTest.php:
```diff
@@ -0,0 +1,74 @@
+        $legacyStockItem->setBackorders(1);
+        $legacyStockItem->setUseConfigBackorders(0);
+        $this->stockItemRepository->save($legacyStockItem);
+        $this->assertFalse($this->isAnySourceInStockCondition->execute('SKU-1', 10));
```
you should check in your test not the particular condition, but the whole service call
`\Magento\InventorySalesApi\Api\IsProductSalableInterface`

maghamed commented on app/code/Magento/InventorySales/Test/Integration/IsProductSalableForRequestedQty/IsAnySourceItemInStockConditionTest.php:
```diff
@@ -0,0 +1,74 @@
+        $legacyStockItem->setBackorders(1);
+        $legacyStockItem->setUseConfigBackorders(0);
+        $this->stockItemRepository->save($legacyStockItem);
+        $this->assertFalse($this->isAnySourceInStockCondition->execute('SKU-1', 10, 1)->isSalable());
```
you should check in your test not the particular condition, but the whole service call
\Magento\InventorySalesApi\Api\IsProductSalableForRequestedQtyInterface::execute

Commit:
MSI-2043: adjusted test and fixed backorders

Commit:
MSI-2043: revert unnesersary changes

Commit:
MSI-2156 Adjust prs #2127, #2201.

novikor committed:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Fixed integration tests

ishakhsuvarov committed:
Merge pull request #2202 from magento-engcom/MSI-2156

MSI-2156 Adjust prs #2127, #2201.

ishakhsuvarov committed:
Merge pull request #2201 from magento-engcom/MSI-2172-adjustements-Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156

Msi 2172 adjustments to Product status is in stock on storefront even it is out of stock in all sources 2156

novikor committed:
Merge branch '2.3-develop' into Product-status-is-In-Stock-on-storefront-even-it-is-Out-of-Stock-in-all-Sources-2156

Commit:
Product status is 'In Stock' on storefront even it is Out of Stock in all Sources #2156.
Fixed case when trying to update source item of product which does not exist anymore

ishakhsuvarov committed:
Merge branch '2.3-develop' of github.com:magento-engcom/msi into 2172-check-for-manage-stock

ishakhsuvarov committed:
#2172: Rework manage stock condition

ishakhsuvarov committed:
Merge pull request #2216 from magento-engcom/2172-check-for-manage-stock

2172 check for manage stock

ghost commented:
Hi @vpodorozh, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento-engcom+msi+2172), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
